shared/domains/chelonia/persistent-actions.js

```diff
@@ -0,0 +1,219 @@
+    // Update relevant status fields before calling any optional code.
+    status.failedAttemptsSoFar++
+    status.lastError = error.message
+    const anyAttemptLeft = options.maxAttempts > status.failedAttemptsSoFar
+    if (!anyAttemptLeft) status.resolved = true
+    status.nextRetry = anyAttemptLeft && !status.resolved
+      ? new Date(Date.now() + options.retrySeconds * 1e3).toISOString()
+      : ''
+    // Perform any optional SBP invocation.
+    sbp('okTurtles.events/emit', PERSISTENT_ACTION_FAILURE, { error, id })
+    await this.trySBP(options.errorInvocation)
+    if (!anyAttemptLeft) {
+      sbp('okTurtles.events/emit', PERSISTENT_ACTION_TOTAL_FAILURE, { error, id })
+      await this.trySBP(options.totalFailureInvocation)
+    }
+    // Schedule a retry if appropriate.
+    if (status.nextRetry) {
+      // Note: there should be no older active timeout to clear.
+      this.timer = setTimeout(() => this.attempt(), this.options.retrySeconds * 1e3)
+    }
+  }
+
+  handleSuccess (result: any): void {
+    const { id, status } = this
+    status.lastError = ''
+    status.nextRetry = ''
+    status.resolved = true
+    sbp('okTurtles.events/emit', PERSISTENT_ACTION_SUCCESS, { id, result })
+  }
+
+  async trySBP (invocation: SbpInvocation | void): Promise<any> {
+    try {
+      return invocation ? await sbp(...invocation) : undefined
+    } catch (error) {
+      console.error(tag, error.message)
+    }
+  }
+}
+
+// SBP API
+
+sbp('sbp/selectors/register', {
+  'chelonia.persistentActions/_init' (): void {
+    this.actionsByID = Object.create(null)
+    // Necessary for now as _init cannot be async. Becomes true when 'configure' has been called.
+    this.ready = false
+    sbp('okTurtles.events/on', PERSISTENT_ACTION_SUCCESS, ({ id }) => {
+      sbp('chelonia.persistentActions/cancel', id)
+    })
+    sbp('okTurtles.events/on', PERSISTENT_ACTION_TOTAL_FAILURE, ({ id }) => {
+      sbp('chelonia.persistentActions/cancel', id)
+    })
+  },
+
+  // Cancels a specific action by its ID.
+  // The action won't be retried again, but an async action cannot be aborted if its promise is stil attempting.
+  'chelonia.persistentActions/cancel' (id: UUIDV4): void {
+    if (id in this.actionsByID) {
+      this.actionsByID[id].cancel()
+      delete this.actionsByID[id]
+      // Likely no need to await this call.
+      sbp('chelonia.persistentActions/save')
+    }
+  },
+
+  // TODO: validation
+  'chelonia.persistentActions/configure' ({ databaseKey, options = {} }: { databaseKey: string; options: Object }): void {
+    if (!databaseKey) throw new TypeError(`${tag} 'databaseKey' is required`)
+    this.databaseKey = databaseKey
+    for (const key in options) {
+      if (key in defaultOptions) {
+        defaultOptions[key] = options[key]
+      } else {
+        throw new TypeError(`${tag} Unknown option: ${key}`)
+      }
+    }
+  },
+
+  'chelonia.persistentActions/enqueue' (...args): UUIDV4[] {
```

Comment:
Good point about the `args` comment.
> Why is it necessary to accept two types of arg for PersistantAction?

Here I just followed the issue requirements.

> And also invocation could be added in options of PersistantAction instance.

An invocation is required in any case, so I don't think `invocation` should go in `options` since it's not actually optional.